### ISSUE TITLE
zoom clips images

### ISSUE THREAD

**Issue by saracarl** [benwbrum/fromthepage#3246]
I am helping to transcribe some documents, and I noticed that when I zoom into the image a lot (to see some detail), the UI clips the edges of the image. This makes it quite difficult to transcribe small text near the edges of the page.

How to reproduce:
Go to https://fromthepage.com/ohiouniversitylibraries/d-day-accounts/cornelius-ryan-wwii-papers-box-008-folder-28-ben-f-pearson/transcribe/32275139
Notice that in the image on the left side, there is some blue vertical text along the right side. 
Use the + button (or scroll wheel) to zoom in about 2-3x.
Pan the view to the right all the way.
Notice that the blue text on the right side is now clipped from the image.

Could you please look into fixing this?

**Comment by saracarl** [comment]
I tried to reproduce the problem described above, but was not able to -- after zooming I had no problem panning to see the vertical text on the far right side of the image.

However, we did have another report of a similar issue.  I note that both documents were imported via IIIF from ContentDM, and their images are served from CDM's IIIF server.  

Compare this image of OpenSeaDragon zoomed to the "home" setting, paying special attention to the right side:
![image](https://github.com/benwbrum/fromthepage/assets/708566/6aef5918-4a49-4e57-b73b-3a609dbc0500)

To the full IIIF image here:
https://cdm17168.contentdm.oclc.org/iiif/2/gvlpiedmont:84/full/full/0/default.jpg
![image](https://github.com/benwbrum/fromthepage/assets/708566/7df6c681-3e77-4a11-a9f9-512b15265792)


Perhaps we should test it with another instance of OSD, if we can find one?  (To rule in or out our layout.)

Here is the page in situ:  https://fromthepage.com/sc-room/greenville-piedmont/1914-01-02-greenville-piedmont/display/32542645


Reported by Shanna at Greenville

**Comment by saracarl** [comment]
#3247 might be the same issue.

**Comment by saracarl** [comment]
#3065 #2733 and #2381 are duplicates of this.

**Comment by rsimon** [comment]
I looked into this and I'm 99% sure that this the image server returning an invalid IIIF response. I've tested in different versions of OpenSeadragon (incl. 3.1 and 4.1, both of which I think are more recent than the version used in FTP); as well as in a IIIF implementation for OpenLayers - all the same result. 

In other words: this beyond the control of FTP, and would have to be addressed by whoever is serving the images. (Might be a matter of upgrading the IIIF server version on their end?)

**Comment by saracarl** [comment]
I don't think it's the image server.  If I drop some of these into Universal Viewer, I can rotate and pan well into the edges with no jumping.   i.e.  https://uv-v4.netlify.app/#?manifest=https://iiif.lib.harvard.edu/manifests/drs:423142543&c=&m=&cv=53&xywh=-174%2C972%2C2981%2C4266

Of the 5 different projects that have documented this, 3 are CONTENTdm hosted (Wisconsin Historical, Ohio University, and Greenville Public Library) and 2 are not (Folger, Harvard).  

The Folger item is not actually IIIF; it was uploaded to FromThePage and is being served as a file.  

This looks old but similar:  https://github.com/openseadragon/openseadragon/issues/469
Possibly related:  https://github.com/openseadragon/openseadragon/issues/469
ohohoh...  This might fix it:  https://github.com/openseadragon/openseadragon/issues/1931#issuecomment-774286328 !!!!

**Comment by rsimon** [comment]
Ah - I see what you mean. I believe these issues talk about different errors.
- The bounce-back issue is indeed caused by the `visibilityRatio` setting
- The original issue at the start of this thread is a different thing, though! I had to look closely myself. But it's not about the view getting "clipped", because the image bounces out. In the above case, the image is actually getting cut off in the view. I'm posting screenshots from the example above:

<img width="246" alt="Bildschirmfoto 2024-01-17 um 07 31 25" src="https://github.com/benwbrum/fromthepage/assets/470971/4ca376a3-a2f9-46e3-a3d7-d1371ef1dd5f">
<img width="246" alt="Bildschirmfoto 2024-01-17 um 07 31 34" src="https://github.com/benwbrum/fromthepage/assets/470971/fd5b0b18-ef2e-484f-80f9-da5e8402ca1f">

The difference between pictures is that I increased the zoom level by one step. If you look closely, you see that the red tab in the second picture is actually missing its left edge. (The handwritten part is chopped off.)

It's is almost a "spot the difference" puzzle ;-) - and it is way more visible if you zoom in and out and see the actual jump happening in the viewer.

The IIIF manifest for this image is this. (UV cannot load it, however, and fails silently, with an error message in the browser console.)
https://cdm15808.contentdm.oclc.org/iiif/2/p15808coll15:12892/info.json

**Comment by saracarl** [comment]
Ah, right.  

I checked[ the newspaper in UV](https://uv-v4.netlify.app/#?manifest=https://cdm17168.contentdm.oclc.org/iiif/gvlpiedmont:88/manifest.json&c=&m=&cv=4&xywh=1828%2C514%2C709%2C495) and it has the same cut-off issue.

I'll pass this on to the CONTENTdm team.

(fyi -- you can find the manifest that was imported by clicking on the work title, then the "about" tab; it's at the bottom of the metadata.)

Closing.

**Comment by saracarl** [comment]
email sent to OCLC:

> Hi OCLC Support,
> 
> Would you pass this on to the CONTENTdm dev team?
> 
> If you open this IIIF manifest (https://cdm17168.contentdm.oclc.org/iiif/gvlpiedmont:88/manifest.json)
>  into either Open SeaDragon (on FromThePage, here:  https://fromthepage.com/sc-room/greenville-piedmont/1914-01-02-greenville-piedmont/transcribe/32542645)
> 
> or Universal Viewer: https://uv-v4.netlify.app/#?manifest=https://cdm17168.contentdm.oclc.org/iiif/gvlpiedmont:88/manifest.json&c=&m=&cv=4&xywh=1859%2C779%2C709%2C495
> 
> Go to page 5 (although any seem to work, the Coca Cola ad makes it most obvious) and zoom about 2 clicks in.  
> 
> You'll notice that the far right of the image gets clipped once you've zoomed; you can't see it or pan to it.
> 
> The behaviour in Mirador is similar but slightly different.  If you load the pages and watch closely, the page initially loads fully, then jumps to a slight zoom (or maybe just jumps) and cuts off the side.  (On further observation, I see this same behavior in FromThePage as well.)
> 
> Given that we see similar behavior on 3 separate IIIF viewers, I think the problem comes from the image server.

**Comment by saracarl** [comment]
Reopening this issue.  OCLC came back and said they thought it was the viewer.
I thought it was the server because 2 separate viewers exhibited the problem, but they are both using open seadragon under the covers.  And the CDM IIIF image URL isn't clipped:  https://cdm17168.contentdm.oclc.org/iiif/2/gvlpiedmont:84/full/max/0/default.jpg

**Comment by rsimon** [comment]
Here's one of the images displayed in Recogito, which uses OpenLayers as the underlying viewer, not OpenSeadragon. Behavior is similarly broken:
https://recogito.pelagios.org/document/dsmybolevh9nqh/part/1/edit

Yes, the default full size image is returned correctly. But the problem is that IIIF viewers request tiles, i.e. sections of the image with given bounds. And this is where their server seems to break. E.g. requests from OpenSedragon, OpenLayers (or any IIIF viewer, really) will look more like this:
https://cdm17168.contentdm.oclc.org/iiif/2/gvlpiedmont:80/0,0,1098,1666/1098,/0/default.jpg



It looks like the returned image doesn't match up with the requested bounds at closer zoom levels. So this must be definitely on their end, I'm afraid.

**Comment by saracarl** [comment]
I think that OpenSeaDragon [depends on OpenLayers](https://github.com/openseadragon/openseadragon/blob/59645e3a0dbc321be0f503df4e6fb4b156c3fbba/src/tmstilesource.js#L48) (although that's the only mention I could find).

UV uses OSD as well.

So...  should we run this through OSD or OpenLayers?  

@rsimon would you be willing to open the issue on whichever one (or both -- one to fix it in OpenLayers and another to get that version into OSD?).

**Comment by rsimon** [comment]
OpenSeadragon and OpenLayers both use their own, independent rendering engines. (The code snippet in question ports a type of zoomable image standard - Tile Maps Service TMS - which was supported in OpenLayers, to OpenSeadragon, so that TMS can also be used with OSD. I should know, I'm the one who ported the code snippet originally years back, as the code comments says ;-) But, importantly: TMS is **not** the protocol that's used in IIIF! I.e. this piece of code never gets traversed when you look at a IIIF image.

That being said: I'm happy to open an issue on the OSD issue tracker about this. Will definitely give us more input & additional opinions. (But I remain 99% convinced that the issue is on the server side ;-)

**Comment by saracarl** [comment]
So here's another tool that uses leaflet-iiif and shows the same issue:
https://ncsu-libraries.github.io/iiif-crop-tool/?newUrl=https%3A%2F%2Fcdm17168.contentdm.oclc.org%2Fiiif%2F2%2Fgvlpiedmont%3A84%2Ffull%2Ffull%2F0%2Fdefault.jpg

Which supports your "it's the image server" problem.

I would have expected leaflet to show the entire image to start with -- I was hoping to use cropping to reproduce it!  I feel like there should be a IIIF image URL that we could pass to the browser (outside of any of the viewer tools) that would show the problem; that would be the simplest reproduction case to pass to Cantaloupe.

**Comment by saracarl** [comment]
So the iiif crop tool is based on [this gist](https://gist.github.com/mejackreed/6936585f435b60aa9451ae2bc1c199f2); all it does is set the tile size:
https://gist.github.com/mejackreed/6936585f435b60aa9451ae2bc1c199f2#file-app-js-L8

Yeah, this needs to go back to CONTENTdm.  It may be a cantaloupe bug, but I don't think we can drive it, they would have to.

**Comment by saracarl** [comment]
Sent back to OCLC support with the following:

Can we reopen this, please?  We've dug in more, and reproduced the problem on 4 separate IIIF viewers that don't use the same tiling libraries:

IIIF Image URL (shows the full image):  https://cdm17168.contentdm.oclc.org/iiif/2/gvlpiedmont:84/full/full/0/default.jpg
Image in Universal Viewer:  https://uv-v4.netlify.app/#?manifest=https://cdm17168.contentdm.oclc.org/iiif/gvlpiedmont:88/manifest.json&c=&m=&cv=4&xywh=1827%2C513%2C709%2C495 
(observe the right side being cut off -- after a default refresh!)
Same Image URL in the IIIF Cropping tool that uses Leaflet:  
https://ncsu-libraries.github.io/iiif-crop-tool/?newUrl=https%3A%2F%2Fcdm17168.contentdm.oclc.org%2Fiiif%2F2%2Fgvlpiedmont%3A84%2Ffull%2Ffull%2F0%2Fdefault.jpg
And in Recogito:  https://recogito.pelagios.org/document/dsmybolevh9nqh/part/1/edit
And if you load the manifest into Mirador (no link, sorry) you'll see the same thing.

That leads us to think it's a problem with the IIIF Image Server (I think y'all use Cantaloupe).

**Comment by rsimon** [comment]
Hi @saracarl, hah - yes, your example with leafleft-iiif provides the perfect evidence, and also the reason for what's going wrong. Thanks for tracking this down.

So... inspecting what happens with Leaflet, here's the perfect reproducible example.
- https://cdm17168.contentdm.oclc.org/iiif/2/gvlpiedmont:84/full/full/0/default.jpg returns the correctly sized, unclipped image right from the CONTENTdm image server.
- If you check the returned image, you can verify that is has an original size of __1186 x 1738__ pixels
- I noticed that leaflet-iiif makes the following initial request: https://cdm17168.contentdm.oclc.org/iiif/2/gvlpiedmont:84/0,0,1138,1668/285,/0/default.jpg This request doesn't match the original image size. Hence, the image server returns a clipped image (as it should, in this case!)
- Looking at the IIIF image manifest, you can see that the __manifest incorrectly reports an image size off1138 x 1668 pixels__. Which is then the root of the problem. https://cdm17168.contentdm.oclc.org/iiif/2/gvlpiedmont:84/info.json

In other words: the CONTENTdm image server is fine. But OCLC are publishing broken manifests that specify an incorrect image size.

**Comment by benwbrum** [comment]
Transferred to OCLC
